### ISSUE TITLE
Fix Ecal top and bottom module edges when gaps are enabled

### ISSUE THREAD

**Issue by ASPatterson** [LDMXAnalysis/ldmx-sw#339]
Hcal wrap inner dimensions were set too tight. They intersected the top and bottom of the Ecal when gap sizes were increased to 2 or more mm. IE the gaps were introduced without checking Hcal intersections.  

Results of intersection test with gaps set to 2mm:
```
-------- WWWW ------- G4Exception-START -------- WWWW -------
*** G4Exception : GeomVol1002
      issued by : G4PVPlacement::CheckOverlaps()
Overlap with volume already placed !
          Overlap is detected for volume em_calorimeters_PV
          with hadronic_calorimeter_PV volume's
          local point (149.6,-257.5,-1592.19), overlapping by at least: 1 mm
NOTE: Reached maximum fixed number -1- of overlaps reports for this volume !
*** This is just a warning message. ***
-------- WWWW -------- G4Exception-END --------- WWWW -------
```

**Particle gun near edge** to illustrate stepping outside sensitive volume. Crashes on second event.
```
# Parse the detector description using the GDML reader.  This assumes that
# there is a local soft link that points to the detector of interest.
/persistency/gdml/read detector.gdml

/ldmx/pw/enable
/ldmx/pw/read spdetector.gdml

/random/setSeeds 5 6
# Enable biasing and set parameters.  This configuration specifies that
# the biasing operator is applied to gammas with energy above 2500 MeV
# that enter the Ecal.  Also, the photonuclear xsec is increased by a factor
# of 1000 within the Ecal only.
/ldmx/biasing/enable
/ldmx/biasing/particle gamma
/ldmx/biasing/process photonNuclear
/ldmx/biasing/volume ecal
/ldmx/biasing/threshold 2500

# Initialize the run
/run/initialize

/ldmx/biasing/xsec/particle gamma
/ldmx/biasing/xsec/process photonNuclear
/ldmx/biasing/xsec/threshold 2500
/ldmx/biasing/xsec/factor 1000

# Position the electron gun upstream of the target.  The initial energy
# of the electron will be 4 GeV.
/gun/particle e-
/gun/energy 4.0 GeV
/gun/position 0 254.9 -0.55 mm
/gun/direction 0 0 4 GeV

# Plugin used to filter events with a hard brem.  In this case, the filter
# is being applied to the target.  Only events where the recoil electron
# has an energy below 1500 MeV are considered.  Also, only events that
# contain at least a single brem gamma above 2500 MeV are considered.
/ldmx/plugins/load TargetBremFilter libBiasing.so
/ldmx/plugins/TargetBremFilter/volume target_PV
/ldmx/plugins/TargetBremFilter/recoil_threshold 1500
/ldmx/plugins/TargetBremFilter/brem_threshold 2500

# Plugin used to filter events where a gamma undergoes a photo-nuclear
# reaction in the target.  In this case, the filter is being applied
# to the Ecal.
/ldmx/plugins/load EcalProcessFilter libBiasing.so
/ldmx/plugins/EcalProcessFilter/volume ecal

# Write the events that pass the filter
/ldmx/persistency/root/verbose 2
/ldmx/persistency/root/file ldmx_sim_events.root

#/tracking/verbose 4
/run/beamOn 10
```
Problem originally appeared as **tracks stepping outside Si near top and bottom edges of Ecal**. Shouldn't have been possible without a volume intersection! Fixed by adjusting the hcal parent volume inner dimensions to allow up to ~ 5 mm gap (well above expected 1-2 mm gap).   

Result of **checking the physical volume** in which Geant thought the hits were in, as they stepped outside the Si (answer: hadronic_calorimeter_PV):
```
>>> Start Event 0 <<<
[EcalHexReadout::getCellModuleID] Someone gave me an out-of-bounds point (-11.259053,257.175769)
NAME OF BAD HIT PHYSICAL VOLUME hadronic_calorimeter_PV
[EcalSD::ProcessHits] Hit step position (-11.259053,257.175769) is outside Ecal Si volume. Trying prePoint position (-10.059884,256.847676) with volume name Si_volume.
[EcalHexReadout::getCellModuleID] Someone gave me an out-of-bounds point (-12.794673,257.001931)
NAME OF BAD HIT PHYSICAL VOLUME em_calorimeters_PV
[EcalSD::ProcessHits] Hit step position (-12.794673,257.001931) is outside Ecal Si volume. Trying prePoint position (-12.458222,257.503863) with volume name Si_volume.
[EcalHexReadout::getCellModuleID] Someone gave me an out-of-bounds point (-12.458222,257.503863)
[EcalHexReadout::getCellModuleID] Someone gave me an out-of-bounds point (-5.324407,257.120360)
NAME OF BAD HIT PHYSICAL VOLUME hadronic_calorimeter_PV
[EcalSD::ProcessHits] Hit step position (-5.324407,257.120360) is outside Ecal Si volume. Trying prePoint position (-5.258832,257.071589) with volume name Si_volume.
[EcalHexReadout::getCellModuleID] Someone gave me an out-of-bounds point (-5.258832,257.071589)
[EcalHexReadout::getCellModuleID] Someone gave me an out-of-bounds point (-5.425651,257.258722)
NAME OF BAD HIT PHYSICAL VOLUME hadronic_calorimeter_PV
[EcalSD::ProcessHits] Hit step position (-5.425651,257.258722) is outside Ecal Si volume. Trying prePoint position (-5.389982,257.169130) with volume name Si_volume.
[EcalHexReadout::getCellModuleID] Someone gave me an out-of-bounds point (-5.389982,257.169130)
[EcalHexReadout::getCellModuleID] Someone gave me an out-of-bounds point (-5.572075,257.483583)
NAME OF BAD HIT PHYSICAL VOLUME hadronic_calorimeter_PV
[EcalSD::ProcessHits] Hit step position (-5.572075,257.483583) is outside Ecal Si volume. Trying prePoint position (-5.461320,257.348314) with volume name Si_volume.
[EcalHexReadout::getCellModuleID] Someone gave me an out-of-bounds point (-5.461320,257.348314)
[EcalHexReadout::getCellModuleID] Someone gave me an out-of-bounds point (-5.921448,257.782134)
NAME OF BAD HIT PHYSICAL VOLUME hadronic_calorimeter_PV
[EcalSD::ProcessHits] Hit step position (-5.921448,257.782134) is outside Ecal Si volume. Trying prePoint position (-5.682829,257.618852) with volume name Si_volume.
[EcalHexReadout::getCellModuleID] Someone gave me an out-of-bounds point (-5.682829,257.618852)
[EcalHexReadout::getCellModuleID] Someone gave me an out-of-bounds point (-6.239001,258.111805)
NAME OF BAD HIT PHYSICAL VOLUME hadronic_calorimeter_PV
[EcalSD::ProcessHits] Hit step position (-6.239001,258.111805) is outside Ecal Si volume. Trying prePoint position (-6.160067,257.945416) with volume name Si_volume.
[EcalHexReadout::getCellModuleID] Someone gave me an out-of-bounds point (-6.160067,257.945416)
[EcalHexReadout::getCellModuleID] Someone gave me an out-of-bounds point (-6.431160,258.090619)
NAME OF BAD HIT PHYSICAL VOLUME hadronic_calorimeter_PV
[EcalSD::ProcessHits] Hit step position (-6.431160,258.090619) is outside Ecal Si volume. Trying prePoint position (-6.317934,258.278195) with volume name Si_volume.
[EcalHexReadout::getCellModuleID] Someone gave me an out-of-bounds point (-6.317934,258.278195)
[EcalHexReadout::getCellModuleID] Someone gave me an out-of-bounds point (-6.455954,257.881056)
NAME OF BAD HIT PHYSICAL VOLUME hadronic_calorimeter_PV
[EcalSD::ProcessHits] Hit step position (-6.455954,257.881056) is outside Ecal Si volume. Trying prePoint position (-6.544385,257.903042) with volume name Si_volume.
[EcalHexReadout::getCellModuleID] Someone gave me an out-of-bounds point (-6.544385,257.903042)
>>> End Event 0 <<<
```
Code which did the above check: Add this into EcalSD just before calling EcalHexReadout (or after, but by default EHR throws an exception for bad points and you'll have to adjust that)
```
// FROM http://geant4.cern.ch/support/faq.shtml#a-geom-1
#include "G4TransportationManager.hh" 
#include "G4Navigator.hh" 
G4VPhysicalVolume* pvol = G4TransportationManager::GetTransportationManager()->GetNavigatorForTracking()->LocateGlobalPointAndSetup(hitPosition); 
std::cout << "NAME OF BAD HIT PHYSICAL VOLUME " << pvol->GetName() << std::endl;
```

**To Test** 
From master, change the gap sizes to 2 mm: A) in DetDescr/include/--/EcalHexReadout.h as defaultGap_{2}, B) in Ecal.gdml as "hexagon_gap" value="2.". Compile. Run the above mac to crash it. Pull in the fix. Run the above mac to see crashes go away. Other choices of the random seeds will make different crashes go away (those numbers just happen to make the 2nd event crash). Also, check that overlaps go away using an overlap mac.

**Comment by ASPatterson** [comment]
![hcalwrapxy_dots](https://user-images.githubusercontent.com/8364791/29113290-7e3520a8-7ca5-11e7-9d63-fa27631c445f.png)
![hcal_z](https://user-images.githubusercontent.com/8364791/29113289-7e17812e-7ca5-11e7-9386-2d047b8b77d0.png)
Hcal hits before fix

**Comment by ASPatterson** [comment]
![hcalxywrap_before](https://user-images.githubusercontent.com/8364791/29114639-c6e3f08c-7ca9-11e7-8bfa-91c63c97a107.png)
![hcal_z_before](https://user-images.githubusercontent.com/8364791/29114640-c6f5cf82-7ca9-11e7-9997-06503b92e290.png)
Hcal hits for gap = 2 mm and below macro, before this fix.
```
# Parse the detector description using the GDML reader.  This assumes that
# there is a local soft link that points to the detector of interest.
/persistency/gdml/read detector.gdml

/ldmx/pw/enable
/ldmx/pw/read spdetector.gdml

/random/setSeeds 5 6
#2 3 on evt 2
# 28 29
# Enable biasing and set parameters.  This configuration specifies that
# the biasing operator is applied to gammas with energy above 2500 MeV
# that enter the Ecal.  Also, the photonuclear xsec is increased by a factor
# of 1000 within the Ecal only.
/ldmx/biasing/enable
/ldmx/biasing/particle gamma
/ldmx/biasing/process photonNuclear
/ldmx/biasing/volume ecal
/ldmx/biasing/threshold 2500

# Initialize the run
/run/initialize

/ldmx/biasing/xsec/particle gamma
/ldmx/biasing/xsec/process photonNuclear
/ldmx/biasing/xsec/threshold 2500
/ldmx/biasing/xsec/factor 1000

# Position the electron gun upstream of the target.  The initial energy
# of the electron will be 4 GeV.
#/gun/particle e-
#/gun/energy 4.0 GeV
#/gun/position 0 0 -0.55 mm
#/gun/position -12.458221 257.503826 -0.55 mm
#/gun/position 0 254.9 -0.55 mm
#/gun/position 3.628210 258.669837 -.55 mm
#/gun/direction 0 0 4 GeV

/ldmx/generators/mpgun/enable
/ldmx/generators/mpgun/nInteractions 1
/ldmx/generators/mpgun/pdgID 11
/ldmx/generators/mpgun/vertex 0 0 -0.55 mm
/ldmx/generators/mpgun/momentum 0 0 4 GeV

/ldmx/generators/beamspot/enable
/ldmx/generators/beamspot/sizeX 600
#; ; 20.0 #total extent 2 x 4 cm
/ldmx/generators/beamspot/sizeY 600
#40.0

# Plugin used to print an event status.
/ldmx/plugins/load EventPrintPlugin

# Plugin used to filter events with a hard brem.  In this case, the filter
# is being applied to the target.  Only events where the recoil electron
# has an energy below 1500 MeV are considered.  Also, only events that
# contain at least a single brem gamma above 2500 MeV are considered.
#/ldmx/plugins/load TargetBremFilter libBiasing.so
#/ldmx/plugins/TargetBremFilter/volume target_PV
#/ldmx/plugins/TargetBremFilter/recoil_threshold 1500
#/ldmx/plugins/TargetBremFilter/brem_threshold 2500
# Plugin used to filter events where a gamma undergoes a photo-nuclear
# reaction in the target.  In this case, the filter is being applied
# to the Ecal.
#/ldmx/plugins/load EcalProcessFilter libBiasing.so
#/ldmx/plugins/EcalProcessFilter/volume ecal

# Print a status every 1000 events.
/ldmx/plugins/EventPrintPlugin/modulus 1000

/ldmx/plugins/load TrackFilterPlugin
/ldmx/plugins/TrackFilterPlugin/verbose 0

#/ldmx/plugins/TrackFilterPlugin/process photonNuclear true
#/ldmx/plugins/TrackFilterPlugin/region CalorimeterRegion true
#/ldmx/plugins/TrackFilterPlugin/create pnFilter

#/ldmx/plugins/TrackFilterPlugin/region CalorimeterRegion true
#/ldmx/plugins/TrackFilterPlugin/create allFilter

# Write the events that pass the filter
/ldmx/persistency/root/verbose 2
/ldmx/persistency/root/file ldmx_sim_events.root

#/tracking/verbose 4

/run/beamOn 1000
```

**Comment by ASPatterson** [comment]
![hcalxywrap_after](https://user-images.githubusercontent.com/8364791/29114700-f1d184b2-7ca9-11e7-99fc-158629b11682.png)
![hcal_z_after](https://user-images.githubusercontent.com/8364791/29114699-f1c6e462-7ca9-11e7-9f9f-8bf6257504fa.png)

And after applying the fix

**Comment by ASPatterson** [comment]
Overlap check afterwards is clean.

**Comment by omar-moreno** [comment]
This issue was resolved by PR 50907a4957229a94a4f11a93668de2601a9deb87.
